crates/tabby-scheduler/src/index.rs

```diff
@@ -39,32 +41,62 @@ pub fn index_repositories(_config: &[RepositoryConfig]) {
-        let text = match file.read_content() {
-            Ok(content) => content,
-            Err(e) => {
-                warn!("Failed to read content of '{}': {}", file.filepath, e);
+    for repository in config {
+        for file in Walk::new(repository.dir()) {
+            let file = match file {
+                Ok(file) => file,
+                Err(e) => {
+                    warn!("Failed to walk file tree for indexing: {e}");
+                    continue;
+                }
+            };
+            let Some(source_file) = cache.get_source_file(repository, file.path()) else {
+                continue;
+            };
+            if !is_valid_file(&source_file) {
+                continue;
+            }
+            let file_id =
+                SourceFileKey::try_from(file.path()).expect("Failed to create source file key");
+
+            if cache.is_indexed(&file_id) {
                 continue;
             }
-        };
-
-        for body in intelligence.chunks(&text) {
-            pb.as_mut()
-                .map(|b| b.update(body.len()))
-                .transpose()
-                .expect("Failed to update progress bar");
-
-            writer
-                .add_document(doc!(
-                        code.field_git_url => file.git_url.clone(),
-                        code.field_filepath => file.filepath.clone(),
-                        code.field_language => file.language.clone(),
-                        code.field_body => body,
-                ))
-                .expect("Failed to add document");
+            let text = match source_file.read_content() {
+                Ok(content) => content,
+                Err(e) => {
+                    warn!(
+                        "Failed to read content of '{}': {}",
+                        source_file.filepath, e
+                    );
+                    continue;
+                }
+            };
+
+            for body in intelligence.chunks(&text) {
+                pb.as_mut()
+                    .map(|b| b.update(body.len()))
+                    .transpose()
+                    .expect("Failed to update progress bar");
+
+                writer
+                    .add_document(doc! {
+                                code.field_git_url => source_file.git_url.clone(),
+                                code.field_file_id => file_id.to_string(),
+                                code.field_filepath => source_file.filepath.clone(),
+                                code.field_language => source_file.language.clone(),
+                                code.field_body => body,
+                    })
+                    .expect("Failed to add document");
+            }
+            cache.set_indexed(&file_id);
         }
     }
 
+    cache.cleanup_old_indexed_files(|key| {
+        writer.delete_term(Term::from_field_text(code.field_file_id, key));
+    });
+
     writer.commit().expect("Failed to commit index");
```

Comment:
write.commit need to be call before the accumulated batch being applied to ensure transactional semantic. (the changes to KV only happens after a successful commitment to tantivy index)